migrating-to-cd4pe.md

```diff
@@ -0,0 +1,126 @@
+
+* Create a pull request from `production` to `staging`
+* Review the changes and merge the PR
+* Create a pull request from `staging` to `development`
+
+In each PR the only changes that should appear in the diff are ones that were not present in the environment above but **not** in the environment below. As long is the logical order was followed, this should never happen. There will however likely be a large number of merge commits. If there are changes, these should be reviewed, merged and deployed using the existing process.
+
+Once all branches have been merged back to the 1st branch (`development`), take a note of the new refs:
+
+| Branch | SHA | Logical Order |
+|--------|-----|---------------|
+| production | `34211f1` | 3rd |
+| staging | `5380862` | 2nd |
+| development | `52c74cb` | 1st |
+
+![Merge-Back](images/data/after_mergeback.png)
+
+### Create the `master` branch
+
+Once the merge-back is complete, create a branch named `master` from the 1st logical branch (`development`). This should contain all of the commits that the current long-lived branches are using. This can be checked using the following command which will return the name of the master branch if it contains the commit.
+
+```shell
+$ git branch --contains 52c74cb
+  development
+* master
+```
+
+![Merge-Back](images/data/master_created.png)
+
+### Create the bookkeeping branches
+
+Now we are ready to create the bookkeeping branches for CD4PE. These are used like moveable tags and are completely controlled by CD4PE once we've started using it, we won't need to mess with them again. Create new branches from the old ones, this essentially duplicates the existing branches. Remember to deploy these branches to the Puppet master.
+
+```shell
+git checkout -b cd4pe_production production
+git push origin cd4pe_production
+git checkout -b cd4pe_staging staging
+git push origin cd4pe_staging
+git checkout -b cd4pe_development development
+git push origin cd4pe_development
+```
+
+### Creating Node Groups
+
+When you deploy in CD4PE you are deploying to a node group. CD4PE does this by taking the branch (environment) that the node group is assigning nodes to, and updating the `HEAD` of that branch to be some new commit. Users much have some method of assigning nodes to environment at the moment, likely it is one of the following ways:
+
+#### Agent-Specified
+
+If agents contain their environment as a setting in `puppet.conf`, and this is not overridden by the master, then new node groups will need to be created to that it *is* overridden by the master using an [environment group](https://puppet.com/docs/pe/2018.1/grouping_and_classifying_nodes.html#create_environment_node_groups). The environment assigned should be the new one we just created i.e. nodes that were previously pointed at the `development` environment should point at the `cd4pe_development` environment.
```

Comment:
```suggestion
If agents contain their environment as a setting in `puppet.conf`, and this is not overridden by the master, then new node groups will need to be created so that it *is* overridden by the master using an [environment group](https://puppet.com/docs/pe/2018.1/grouping_and_classifying_nodes.html#create_environment_node_groups). The environment assigned should be the new one we just created i.e. nodes that were previously pointed at the `development` environment should point at the `cd4pe_development` environment.
```